### PR TITLE
#2858 - fixed topics same slug tag does not add issue

### LINKED ISSUE #2858: New tag does not add when tag exists with same slug
**Describe the bug**
 new tag does not add when tag exists with same slug https://i.imgur.com/sU8RO4b.png

**To Reproduce**
1. Create a discussion topic with tag like "@c#"
2. Edit the topic and add new tag like "@c++"
3. Save it and check the tags 
4. See that "@c++" does not add

**Expected behavior**
New tags with the same slug should be added

**Screenshots**
https://i.imgur.com/sU8RO4b.png

**Related platform api issue**
https://github.com/buddyboss/buddyboss-platform-api/issues/326

**Related app native issue for QA**
https://github.com/buddyboss/buddyboss-app-native/issues/1598

### PR DESCRIPTION
**Pull request by sadrul**
**Related platform api PR**
https://github.com/buddyboss/buddyboss-platform-api/pull/329

**Related app native issue for QA**
buddyboss/buddyboss-app-native#1598

### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Fixes #2858  .

### How to test the changes in this Pull Request:

1. Create a discussion topic with tag like "@c#"
2. Edit the topic and add new tag like "@c++"
3. Save it and check the tags 
4. See that "@c++" has been added now

please check for **create topic, edit topic, create reply, edit reply**

### Proof Screenshots or Video
https://www.loom.com/share/49b75a50486640cc826330211040c7b2

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### CODE REVIEW

File: src/bp-forums/replies/functions.php [hunk 1]
```diff
@@ -433,7 +433,33 @@ function bbp_new_reply_handler( $action = '' ) {
...
+				if ( ! $term_info ) {
+					$term_info = wp_insert_term( $term_name, bbp_get_topic_tag_tax_id(), $args );
+				}
+				if ( ! empty( $term_info ) && ! is_wp_error( $term_info ) ) {
+					$term_ids[] = $term_info['term_id'];
+				}
+			}
+
+			if ( ! empty( $term_ids ) ) {
+				$terms = bp_set_object_terms( $topic_id, $term_ids, bbp_get_topic_tag_tax_id(), true );
```
Comment (KartikSuthar) [inline src/bp-forums/replies/functions.php:460]:
```suggestion
				$terms = wp_set_post_terms( $topic_id, $term_ids, bbp_get_topic_tag_tax_id(), true );
```

File: src/bp-forums/replies/functions.php [hunk 2]
```diff
@@ -785,23 +811,53 @@ function bbp_edit_reply_handler( $action = '' ) {
...
+			if ( ! $term_info ) {
+				$term_info = wp_insert_term( $term_name, bbp_get_topic_tag_tax_id(), $args );
+			}
+			if ( ! empty( $term_info ) && ! is_wp_error( $term_info ) ) {
+				$term_ids[] = $term_info['term_id'];
+			}
+		}
+
+		if ( ! empty( $term_ids ) ) {
+			$terms = bp_set_object_terms( $topic_id, $term_ids, bbp_get_topic_tag_tax_id(), true );
```
Comment (KartikSuthar) [inline src/bp-forums/replies/functions.php:858]:
```suggestion
			$terms = wp_set_post_terms( $topic_id, $term_ids, bbp_get_topic_tag_tax_id(), true );
```

File: src/bp-forums/topics/functions.php [hunk 3]
```diff
@@ -390,6 +389,30 @@ function bbp_new_topic_handler( $action = '' ) {
...
+				if ( ! $term_info ) {
+					$term_info = wp_insert_term( $term_name, bbp_get_topic_tag_tax_id(), $args );
+				}
+				if ( ! empty( $term_info ) && ! is_wp_error( $term_info ) ) {
+					$term_ids[] = $term_info['term_id'];
+				}
+			}
+
+			if ( ! empty( $term_ids ) ) {
+				bp_set_object_terms( $topic_id, $term_ids, bbp_get_topic_tag_tax_id(), true );
```
Comment (KartikSuthar) [inline src/bp-forums/topics/functions.php:412]:
```suggestion
				wp_set_post_terms( $topic_id, $term_ids, bbp_get_topic_tag_tax_id(), true );
```

File: src/bp-forums/topics/functions.php [hunk 4]
```diff
@@ -767,6 +789,52 @@ function bbp_edit_topic_handler( $action = '' ) {
...
```
Comment (KartikSuthar) [inline src/bp-forums/topics/functions.php:834]:
```suggestion
				wp_set_post_terms( $topic_id, $term_ids, bbp_get_topic_tag_tax_id(), true );
```

**Review by KartikSuthar — CHANGES_REQUESTED** [review]

File: src/bp-forums/replies/functions.php [hunk 5]
```diff
@@ -433,7 +433,13 @@ function bbp_new_reply_handler( $action = '' ) {
...
 
 		// Insert terms.
-		$terms = wp_set_post_terms( $topic_id, $terms, bbp_get_topic_tag_tax_id(), true );
+		$topic_taxonomy = bbp_get_topic_tag_tax_id();
+		if ( ! is_array( $terms ) && strstr( $terms, ',' ) ) {
+			$terms = explode( ',', $terms );
+		} else {
+			$terms = (array) $terms;
+		}
+		$terms = bb_add_topic_tags( $terms, $topic_id, $topic_taxonomy );
```
Comment (chetansatasiya) [inline src/bp-forums/replies/functions.php:442]:
```suggestion
		if ( ! is_array( $terms ) && strstr( $terms, ',' ) ) {
			$terms = explode( ',', $terms );
		} else {
			$terms = (array) $terms;
		}
		$terms = bb_add_topic_tags( $terms, $topic_id, bbp_get_topic_tag_tax_id() );
```

File: src/bp-forums/replies/functions.php [hunk 6]
```diff
@@ -781,27 +787,15 @@ function bbp_edit_reply_handler( $action = '' ) {
...
-
-		if ( ! empty( $terms ) ) {
-			$terms_array   = explode( ',', $terms );
-			$deleted_terms = array_diff( $existing_terms, $terms_array );
-		}
-		if ( ! empty( $deleted_terms ) ) {
-			wp_remove_object_terms( $topic_id, $deleted_terms, bbp_get_topic_tag_tax_id() );
-		}
+	// update terms
+	$topic_taxonomy = bbp_get_topic_tag_tax_id();
```
Comment (chetansatasiya) [inline src/bp-forums/replies/functions.php:791]:
```suggestion
```

File: src/bp-forums/replies/functions.php [hunk 7]
```diff
@@ -781,27 +787,15 @@ function bbp_edit_reply_handler( $action = '' ) {
...
+	if ( ! is_array( $terms ) && strstr( $terms, ',' ) ) {
+		$terms = explode( ',', $terms );
+	} else {
+		$terms = (array) $terms;
 	}
-
-	// Insert terms
-	$terms = wp_set_post_terms( $topic_id, $terms, bbp_get_topic_tag_tax_id(), true );
+	$existing_terms = bbp_get_topic_tag_names( $topic_id );
+	$terms = bb_add_topic_tags( $terms, $topic_id, $topic_taxonomy, $existing_terms );
```
Comment (chetansatasiya) [inline src/bp-forums/replies/functions.php:798]:
```suggestion
	$terms = bb_add_topic_tags( $terms, $topic_id, bbp_get_topic_tag_tax_id(), $existing_terms );
```

File: src/bp-forums/topics/functions.php [hunk 8]
```diff
@@ -767,6 +769,11 @@ function bbp_edit_topic_handler( $action = '' ) {
 
 	if ( ! empty( $topic_id ) && ! is_wp_error( $topic_id ) ) {
 
+		// update tags.
+		$topic_taxonomy = bbp_get_topic_tag_tax_id();
+		$existing_terms = bbp_get_topic_tag_names( $topic_id );
+		bb_add_topic_tags( (array) $terms[ $topic_taxonomy ], $topic_id, $topic_taxonomy, $existing_terms );
+
```
Comment (chetansatasiya) [inline src/bp-forums/topics/functions.php:776]:
```suggestion
		bb_add_topic_tags( (array) $terms[ $topic_taxonomy ], $topic_id, bbp_get_topic_tag_tax_id(), bbp_get_topic_tag_names( $topic_id ) );

```

**Review by chetansatasiya — CHANGES_REQUESTED** [review]
Need changes

**Review by chetansatasiya — DISMISSED** [review]

**Comment by sadrul** [comment]
@sumitbuddyboss @mohit-buddyboss merged into dev and qa-ongoing and ready for testing

**Comment by mohit-buddyboss** [comment]
@sadrul 

Tested on qa-ongoing branch of the platform.
Now we are able to create tags having same slug and also edit/delete functionality is working fine.
Video: https://loom.com/share/6ee49a4af48741ee8ff7b4a3165e6bc4

Related issue is resolve now and PR is ready for release.

cc @sourabhmatolia @firozwp @KartikSuthar @chetansatasiya

**Comment by JetalDobariya** [comment]
Hello @sadrul 

I did verify this PR on qa-approved branch, now tags are working fine when we add tags like "C , C++ , C#".
--> https://www.loom.com/i/c89f1887ad2c4d659d22ee781116b72a

Thanks!

**Pull request merged**